cmd/zed/agents/zfs_mod.c

```diff
@@ -592,11 +591,26 @@ zfs_iter_vdev(zpool_handle_t *zhp, nvlist_t *nvl, void *data)
 		 * illumos, substring matching is not required to accommodate
 		 * the partition suffix. An exact match will be present in
 		 * the dp->dd_compare value.
+		 * If the attached disk already contains a vdev GUID, it means
+		 * the disk is not clean. In such a scenario, the physical path
+		 * would be a match that makes the disk faulted when trying to
+		 * online it. So, we would only want to proceed if either GUID
+		 * matches with the last attached disk or the disk is in clean
+		 * state.
 		 */
 		if (nvlist_lookup_string(nvl, dp->dd_prop, &path) != 0 ||
-		    strcmp(dp->dd_compare, path) != 0) {
-			zed_log_msg(LOG_INFO, "  %s: no match (%s != vdev %s)",
-			    __func__, dp->dd_compare, path);
+		    strcmp(dp->dd_compare, path) != 0 || (dp->dd_new_vdev_guid
+		    != 0 && dp->dd_new_vdev_guid != guid)) {
+			if (dp->dd_new_vdev_guid != 0) {
+				zed_log_msg(LOG_INFO, "  %s: no match (%s,%llu"
+				    " != vdev %s,%llu)", __func__,
+				    dp->dd_compare, dp->dd_new_vdev_guid, path,
+				    guid);
+			} else {
+				zed_log_msg(LOG_INFO, "  %s: no match (%s != "
+				    "vdev %s)", __func__, dp->dd_compare,
+				    path);
+			}
```

Comment:
The code would be much more readable if you split this block into two, one checking paths as-is, and add another separate matching the guids.  Then you would not need second dp->dd_new_vdev_guid != 0 check (that is actually incorrect if we got here due to path mismatch).